pipeline/exec/undo_manager.go

```diff
@@ -0,0 +1,142 @@
+package exec
+
+import (
+	"context"
+	"errors"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/streamingfast/bstream"
+	pbbstream "github.com/streamingfast/bstream/pb/sf/bstream/v1"
+	"go.uber.org/atomic"
+	"go.uber.org/zap"
+)
+
+var GlobalUndoManager *UndoManager
+var ErrBlockUndo = errors.New("block undone")
+var MaxUndoneBlocksSize = 100
+
+var globalIDCounter atomic.Uint64
+
+func nextUniqueID() uint64 {
+	return globalIDCounter.Add(1)
+}
```

Comment:
Tie it to `UndoManager` struct instead of being a global, id needs to be unique per manager anyway.